File: 295-find-median-from-data-stream/295-find-median-from-data-stream.cpp
class MedianFinder {
    priority_queue<int> maxHeap;
    priority_queue<int,vector<int>,greater<int>> minHeap;
public:
    int size;
    MedianFinder() {
        size = 0;
    }
    
    void addNum(int num) {
        if(minHeap.empty() || minHeap.top() <= num)
            minHeap.push(num);
        else
            maxHeap.push(num);
        ++size;
        if(minHeap.size() > maxHeap.size() + size%2)
        {
            maxHeap.push(minHeap.top());
            minHeap.pop();
        }
        else if(maxHeap.size() > minHeap.size())
        {
            minHeap.push(maxHeap.top());
            maxHeap.pop();
        }
    }
    
    double findMedian() {
        if( size % 2 != 0 )
            return minHeap.top();
        return (minHeap.top() + maxHeap.top())/2.0;        
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 295-find-median-from-data-stream/295-find-median-from-data-stream.java
class MedianFinder {
    PriorityQueue<Integer> maxHeap = new PriorityQueue(Collections.reverseOrder());
    PriorityQueue<Integer> minHeap = new PriorityQueue();
    
    public MedianFinder() {
        
    }
    
    public void addNum(int num) {
        if (maxHeap.isEmpty() || maxHeap.peek() >= num ) {
            maxHeap.add(num);
        } else {
            minHeap.add(num);
        }
        
        if(minHeap.size()+1 < maxHeap.size()) {
            minHeap.add(maxHeap.poll());
            
        }else if(minHeap.size() > maxHeap.size()) {
            maxHeap.add(minHeap.poll());
        }
    }
    
    public double findMedian() {
        if(maxHeap.size() == minHeap.size()) {
            return Double.valueOf((minHeap.peek()+maxHeap.peek())/2d);
        } 
        return Double.valueOf(maxHeap.peek());
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */